File: cpp-solutions/medium/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if (s.length() == 0) return 0;
        int chars[128] = {0};
        int ans = 0, left = 0, right = 0;
        while (right < s.length()) {
            char r = s[right];
            chars[r]++; //indexed

            while (chars[r] > 1) {
                char l = s[left];
                chars[l]--;
                left++;
            }

            ans = max(ans, right - left + 1);
            right++;
        }
        return ans;
    }
};

File: python3-solutions/medium/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) == 0 : return 0
        chars = [0] * 128
        ans, left, right = 0, 0, 0
        while right < len(s):
            r = s[right]
            chars[ord(r)] = chars[ord(r)]+1

            while chars[ord(r)] > 1 :
                l = s[left]
                chars[ord(l)] = chars[ord(l)]-1
                left = left+1

            ans = max(ans, right - left + 1)
            right = right+1
        return ans